pkg/driver/driver_openstack.go

```diff
@@ -0,0 +1,213 @@
+
+	return result.Err
+}
+
+// GetExisting TODO
+func (d *OpenStackDriver) GetExisting() (string, error) {
+	return d.MachineId, nil
+}
+
+// createNovaClient TODO
+func (d *OpenStackDriver) createNovaClient() (*gophercloud.ServiceClient, error) {
+
+	config := &tls.Config{}
+
+	authURL, ok := d.CloudConfig.Data["authURL"]
+	if !ok {
+		return nil, fmt.Errorf("missing auth_url in secret")
+	}
+	glog.Infof("AuthURL: %s", authURL)
+	username, ok := d.CloudConfig.Data["username"]
+	if !ok {
+		return nil, fmt.Errorf("missing username in secret")
+	}
+	glog.Infof("Username: %s", username)
+	password, ok := d.CloudConfig.Data["password"]
+	if !ok {
+		return nil, fmt.Errorf("missing password in secret")
+	}
+	domainName, ok := d.CloudConfig.Data["domainName"]
+	if !ok {
+		return nil, fmt.Errorf("missing domainName in secret")
+	}
+	glog.Infof("DomainName: %s", domainName)
+	region := d.OpenStackMachineClass.Spec.Region
+	glog.Infof("Region: %s", region)
+	tenantName, ok := d.CloudConfig.Data["tenantName"]
+	if !ok {
+		return nil, fmt.Errorf("missing tenantName in secret")
+	}
+	glog.Infof("TenantName: %s", tenantName)
+
+	caCert, ok := d.CloudConfig.Data["caCert"]
+	if !ok {
+		caCert = nil
+	}
+
+	insecure, ok := d.CloudConfig.Data["insecure"]
+	if ok && strings.TrimSpace(string(insecure)) == "true" {
+		config.InsecureSkipVerify = true
+	}
+
+	if caCert != nil {
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM([]byte(caCert))
+		config.RootCAs = caCertPool
+	}
+
+	clientCert, ok := d.CloudConfig.Data["clientCert"]
+	if ok {
+		clientKey, ok := d.CloudConfig.Data["clientKey"]
+		if ok {
+			cert, err := tls.X509KeyPair([]byte(clientCert), []byte(clientKey))
+			if err != nil {
+				return nil, err
+			}
+			config.Certificates = []tls.Certificate{cert}
+			config.BuildNameToCertificate()
+		} else {
+			return nil, fmt.Errorf("client key missing in secret")
+		}
+	}
+
+	opts := gophercloud.AuthOptions{
+		IdentityEndpoint: strings.TrimSpace(string(authURL)),
+		Username:         strings.TrimSpace(string(username)),
+		Password:         strings.TrimSpace(string(password)),
+		DomainName:       strings.TrimSpace(string(domainName)),
+		TenantName:       strings.TrimSpace(string(tenantName)),
+	}
+	glog.Infof("Auth opts")
+
+	client, err := openstack.NewClient(opts.IdentityEndpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set UserAgent
+	client.UserAgent.Prepend("Machine Controller 08/15")
+
+	transport := &http.Transport{Proxy: http.ProxyFromEnvironment, TLSClientConfig: config}
+	client.HTTPClient = http.Client{
+		Transport: transport,
+	}
+
+	err = openstack.Authenticate(client, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return openstack.NewComputeV2(client, gophercloud.EndpointOpts{
+		Region:       strings.TrimSpace(region),
+		Availability: gophercloud.AvailabilityPublic,
+	})
+}
+
+func (d *OpenStackDriver) encodeMachineId(instanceId string) string {
+	return fmt.Sprintf("openstack:///%s \n", instanceId)
```

Comment:
Add region of the machine in instanceID on creation, similar to the case of AWS driver. Also please remove the trailing '\n' in the encode function.